src/drivers/snapdragon_pwm_out/snapdragon_pwm_out.cpp

```diff
@@ -0,0 +1,608 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <stdint.h>
+
+#include <px4_tasks.h>
+#include <px4_getopt.h>
+#include <px4_posix.h>
+#include <errno.h>
+#include <cmath>	// NAN
+
+
+#include <uORB/uORB.h>
+#include <uORB/topics/actuator_controls.h>
+#include <uORB/topics/actuator_outputs.h>
+#include <uORB/topics/actuator_armed.h>
+
+#include <drivers/drv_hrt.h>
+#include <drivers/drv_mixer.h>
+#include <systemlib/mixer/mixer.h>
+#include <systemlib/mixer/mixer_load.h>
+#include <systemlib/mixer/mixer_multirotor.generated.h>
+#include <systemlib/param/param.h>
+#include <systemlib/pwm_limit/pwm_limit.h>
+#include <dev_fs_lib_pwm.h>
+
+/*
+ * This driver is supposed to run on Snapdragon.
+ * Author: Dennis Mannhart
+ * email: dennis.mannhart@gmail.com
+ */
+namespace snapdragon_pwm
+{
+
+static px4_task_t _task_handle = -1; // handle to the task main thread
+volatile bool _task_should_exit = false; // should task exit
+static bool _is_running = false; // is app running
+
+static const int NUM_PWM = 4;
+static char _device[64] = "/dev/pwm-1";
+
+// pwm configuration with start pin 27
+static const int PIN_GPIO = 27;
+static struct ::dspal_pwm_ioctl_signal_definition _signal_definition;
+static struct ::dspal_pwm _pwm_gpio[NUM_PWM];
+static struct ::dspal_pwm_ioctl_update_buffer *_update_buffer;
+static struct ::dspal_pwm *_pwm;
+
+
+// ToDo: check frequency required
+static const int FREQUENCY_PWM = 400;
+
+
+/* filename: /dev/fs is mapped to /usr/share/data/adsp */
+static const char *MIXER_FILENAME = "/dev/fs/quad_x.main.mix";
+
+// subscriptions
+int	_controls_subs[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+int	_armed_sub;
+int _fd = -1;
+
+// publications
+orb_advert_t    _outputs_pub = nullptr;
+
+// topic structures
+actuator_controls_s _controls[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+orb_id_t _controls_topics[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+actuator_outputs_s  _outputs;
+actuator_armed_s    _armed;
+
+// polling
+uint8_t _poll_fds_num = 0;
+px4_pollfd_struct_t _poll_fds[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+
+// control group
+uint32_t	_groups_required = 0;
+
+// limit for pwm
+pwm_limit_t     _pwm_limit;
+
+// esc parameters
+int32_t _pwm_disarmed;
+int32_t _pwm_min;
+int32_t _pwm_max;
+
+MultirotorMixer *_mixer = nullptr;
+
+
+/*
+ * forward declaration
+ */
+
+void usage();
+
+void start();
+
+void stop();
+
+int pwm_initialize(const char *device);
+
+void pwm_deinitialize();
+
+void send_outputs_pwm(const uint16_t *pwm);
+
+void task_main_trampoline(int argc, char *argv[]);
+
+void subscribe();
+
+void task_main(int argc, char *argv[]);
+
+int initialize_mixer(const char *mixer_filename);
+
+int mixer_control_callback(uintptr_t handle, uint8_t control_group, uint8_t control_index, float &input);
+
+
+
+/*
+ * functions
+ */
+
+int mixer_control_callback(uintptr_t handle,
+			   uint8_t control_group,
```

Comment:
One space more :). It's probably easiest to make sure the file gets checked with `make format`.